route parameters can also be other things than string

dbu opened the pull request:
fix #240

dbu commented on src/Knp/Menu/Matcher/Voter/RouteVoter.php:
```diff
@@ -78,7 +78,8 @@ private function isMatchingRoute(array $testedRoute)
         foreach ($testedRoute['parameters'] as $name => $value) {
-            if (!isset($routeParameters[$name]) || $routeParameters[$name] !== (string) $value) {
+            // use a weak comparison, parameters can be numeric or strings
+            if (!isset($routeParameters[$name]) || $routeParameters[$name] != $value) {
```
@stof can you agree with this? or should i cast both sides to string?

stof replied:
using a loose comparison is a bad idea, because ``0 == 'abc'`` will be true.

dbu commented:
@greg0ire so this "fix" actually shows a problem. there is one test that we break with this change, which works with integer parameters. can it be that you misconfigured your menu item with a string instead of a number?

greg0ire commented:
Not sure how I can check this, but I'm using the identifier of a doctrine entity, which is an integer.

greg0ire commented:
If I re-read @stof comments, I think casting both sides to string would be the solution : 
- when the id comes from a query string parameter, it is always a string, and this is what the failing unit tests was written for
- when the id comes from a default value, it can be an integer, and this is what a test should be written for.

greg0ire commented:
I'm trying to simplify these tests before in #244

stof commented:
there is a discussion in Symfony currently to decide whether we should deprecate non-string default values. So I suggest waiting for the outcome of the discussion before breaking BC here (because changing it can break BC).

dbu commented:
@stof do you know the outcome of the symfony discussion?

dbu commented:
@stof do you know the outcome of the symfony discussion?

dbu commented:
ping @stof

dbu commented:
so should i rebase this PR? and should i go for a strict comparison but just remove casting to string?

stof commented:
Simple strict comparison is a no-go, as it would break things for people using `$entity->getId()` when building the menu (which will be an integer) to compare it against the matched id (which will be a string).
Loose comparison is a no-go due to WTF comparison for int vs string (casting to int is lossy).

so I would go with the option of casting the second side to string as well.

stof commented:
@dbu do you have an idea about when you will be able to finish this one ? I'm wondering whether I should wait for it to release 2.3 (I'd like to do it tomorrow), or let it be part of a later patch release.

dbu commented:
i'll have to look into the build error. looks to me like the test is assuming the incorrect behaviour. but i have meetings now, might only get to it towards the evening.

should i also add a note to the changelog? its a bugfix but changes behaviour and people might have relied on this bug...

stof commented:
@dbu you might add a note in the changelog. Otherwise I'll do it myself before the release (many other PRs are missing the changelog entries, including mines)

dbu commented:
updated the test, i think it should indeed change behaviour here. and added a changelog note. (i am not only anymore tonight, feel free to adjust this to your requirements if you want tweaks and want to release). given that this is a bugfix-behaviour-change, it would seem nicer to me to have it in a new minor version than in a patch release.
